server/src/main/java/org/elasticsearch/common/logging/ThrottlingAndHeaderWarningLogger.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.LogManager;
+
+/**
+ * This class wraps both <code>HeaderWarningLogger</code> and <code>ThrottlingLogger</code>
+ * which is a common use case across Elasticsearch
+ */
+public class ThrottlingAndHeaderWarningLogger {
+    private final HeaderWarningLogger headerWarningLogger = new HeaderWarningLogger();
+    private final ThrottlingLogger throttlingLogger;
+
+    public ThrottlingAndHeaderWarningLogger(String loggerName) {
+        this.throttlingLogger = new ThrottlingLogger(LogManager.getLogger(loggerName));
+    }
+
+    /**
+     * Logs a message, adding a formatted warning message as a response header on the thread context.
+     */
+    public void headerWarnAndLog(String msg, Object... params) {
```

Comment:
can we make this just `logAndWarn` I think that is different enough from `info`, `warn` etc. and you can always look up at the type of logger.

Reply:
can we also expose a `log` and a `warn` method ?  log = logfile, warn = header

Reply:
_ninja edit_ - s/warn/header (in case the comments here don't line up with the contents in the email updates)

Reply:
agree - log and header method will be useful too. Not sure about the header method name though https://github.com/elastic/elasticsearch/pull/55941#discussion_r420695963

Reply:
hm at the same time there is no usage for just log/header on its own at the moment.
maybe let's keep that class small for now and add methods as the need arise?